### PR TITLE
[Diagnostics] Port diagnostics for array/dictionary literals

### PR DESCRIPTION
**Pull request by xedin**
`CollectionElementContextualMismatch` covers almost all of the related diagnostics
already with a single exception - generic argument mismatches which has been "fixed"
already just didn't have a diagnostic support.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSDiagnostics.cpp [hunk 1]
```diff
@@ -670,6 +670,18 @@ bool GenericArgumentsMismatchFailure::diagnoseAsError() {
       break;
     }
 
+    case ConstraintLocator::TupleElement: {
+      if (auto *array = dyn_cast<ArrayExpr>(getRawAnchor()))
```
Comment (theblixguy) [inline lib/Sema/CSDiagnostics.cpp]: Since `array` is not used, perhaps you could just use an `isa`?

File: lib/Sema/CSDiagnostics.cpp [hunk 2]
```diff
@@ -670,6 +670,18 @@ bool GenericArgumentsMismatchFailure::diagnoseAsError() {
...
+        diagnostic = getDiagnosticFor(CTP_ArrayElement);
+
+      if (auto *dict = dyn_cast<DictionaryExpr>(getRawAnchor())) {
```
Comment (theblixguy) [inline lib/Sema/CSDiagnostics.cpp]: Same here - `dict` is unused.

File: lib/Sema/CSDiagnostics.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Yeah, make sense. Thanks!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]
Awesome!

**Pull request merged**